tracer/src/Datadog.Trace/Telemetry/TelemetrySettings.cs

```diff
@@ -13,16 +12,37 @@ internal class TelemetrySettings
     {
         public TelemetrySettings(IConfigurationSource source, ImmutableTracerSettings tracerSettings)
         {
+            var apiKey = source?.GetString(ConfigurationKeys.ApiKey);
             var explicitlyEnabled = source?.GetBool(ConfigurationKeys.Telemetry.Enabled);
-            TelemetryEnabled = explicitlyEnabled ?? false;
+            var agentlessExplicitlyEnabled = source?.GetBool(ConfigurationKeys.Telemetry.AgentlessEnabled);
+            bool agentlessEnabled = false;
 
-            var apiKey = source?.GetString(ConfigurationKeys.ApiKey);
+            if (agentlessExplicitlyEnabled == true)
+            {
+                if (string.IsNullOrEmpty(apiKey))
+                {
+                    ConfigurationError = "Telemetry configuration error: Agentless mode was enabled, but no API key was available";
+                }
+                else
+                {
+                    agentlessEnabled = true;
+                }
+            }
+            else if (agentlessExplicitlyEnabled is null)
+            {
+                // if there's an API key, we use agentless mode, otherwise we use the agent
```

Comment:
Is this behavior documented somewhere as well? for other tracers to follow.

Reply:
Are we sure we want to make this decision based on the presence of the API key? Users can have `DD_API_KEY` set even if the Agent is present. (I do this on my own machine, for example.) Should we prefer using the Agent in that case? 

I'm wondering if we should default to using the Agent unless the user explicitly enables Agentless mode. A similar discussion came up recently when talking about Agentless for traces.

Reply:
> Are we sure we want to make this decision based on the presence of the API key

No, we're not sure 😅 

> Users can have DD_API_KEY set even if the Agent is present... Should we prefer using the Agent in that case

This is the crux of the matter. There's several reasons for this decision:

- _Currently_, very few people will have an agent that supports sending telemetry, as it's only supported in the latest version of the agent.
- We don't (yet) support _changing_ configuration after the initial setup. So that means we _can't_ try and send to the agent, and if that fails, fallback to agentless.
- We _could_ call the `/info` endpoint of the agent, to find out if supports telemetry. But that means we can't finish setup until _after_ making an HTTP call. This configuration is currently in the application (and Tracer) startup hot-path, so we can't do that as-is.
- Given all of the above, defaulting to using the agent will be the "wrong" option in most cases where we have the api key. There's (currently) not many reasons _not_ to use the api key if it's available.

These tradeoffs are different concerns to using agentless mode for traces. The performance profile will be significantly different between agentless and agent (agent-ed? agent-full?) tracing, which isn't the case for telemetry. For telemetry we do the same work either way, and just send the data to one place or another.

So, tl;dr;

- I think this behaviour is correct for now
- Before we enable telemetry by default, we should consider refactoring the configuration to be "lazy", so that we can check the agent version before deciding which approach to use. I'd like to do this in a separate PR if possible though

Reply:
> I think this behaviour is correct for now

Makes sense, thank you!

> we should consider refactoring the configuration to be "lazy", so that we can check the agent version before deciding which approach to use

Yeah, we will need this for other features as well.